[HLSL] Implement array temporary support

llvm-beanz opened the pull request:
HLSL constant sized array function parameters do not decay to pointers.
Instead constant sized array types are preserved as unique types for
overload resolution, template instantiation and name mangling.

This implements the change by adding a new `ArrayParameterType` which
represents a non-decaying `ConstantArrayType`. The new type behaves the
same as `ConstantArrayType` except that it does not decay to a pointer.

Values of `ConstantArrayType` in HLSL decay during overload resolution
via a new `HLSLArrayRValue` cast to `ArrayParameterType`.

`ArrayParamterType` values are passed indirectly by-value to functions
in IR generation resulting in callee generated memcpy instructions.

The behavior of HLSL function calls is documented in the [draft language specification](https://microsoft.github.io/hlsl-specs/specs/hlsl.pdf) under the Expr.Post.Call heading.

Additionally the design of this implementation approach is documented in [Clang's documentation](https://clang.llvm.org/docs/HLSL/FunctionCalls.html)

Resolves #70123

llvm-beanz commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -10524,6 +10524,11 @@ Sema::PerformCopyInitialization(const InitializedEntity &Entity,
+  if (LangOpts.HLSL)
+    if (auto AdjTy = dyn_cast<DecayedType>(Entity.getType()))
+      if (AdjTy->getOriginalType()->isConstantArrayType())
+        InitE = HLSLArrayTemporaryExpr::Create(getASTContext(), InitE);
```
Ooo. I like this idea a lot. Let me go give that a try. I was trying to avoid completely breaking array decay completely because at some point we’re going to align the language more with C/C++.

llvm-beanz committed:
[HLSL] Pass arrays by value

HLSL constant sized array function parameters do not decay to pointers.
Instead constant sized array types are preserved as unique types for
overload resolution, template instantiation and name mangling.

This implements the change by adding a new `ArrayParameterType` which
represents a non-decaying `ConstantArrayType`. The new type behaves the
same as `ConstantArrayType` except that it does not decay to a pointer.

Values of `ConstantArrayType` in HLSL decay during overload resolution
via a new `HLSLArrayRValue` cast to `ArrayParameterType`.

`ArrayParamterType` values are passed indirectly by-value to functions
in IR generation resulting in callee generated memcpy instructions.

llvm-beanz commented:
I force pushed an update here because the new implementation is basically a complete rewrite based on @efriedma-quic's suggestion. I think this is a significantly cleaner approach that more accurately models the behavior of HLSL.

llvm-beanz committed:
Update docs based on the implementation that I posted rather than an
earlier iteration.

llvm-beanz committed:
One more doc update

efriedma-quic commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -10900,6 +10936,10 @@ QualType ASTContext::mergeTypes(QualType LHS, QualType RHS, bool OfBlockPointer,
            "Equivalent pipe types should have already been handled!");
     return {};
+  case Type::ArrayParameter:
+    if (RHS != LHS)
```
It's impossible for RHS and LHS to be equal here.

efriedma-quic commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -4655,6 +4655,13 @@ void CodeGenFunction::EmitCallArg(CallArgList &args, const Expr *E,
+  // If an argument is an array paramter expression being passed through. Emit
+  // the argument to a temporary and pass the temporary as the call arg.
+  if (auto AT = dyn_cast<ArrayParameterType>(type)) {
+    args.add(EmitAnyExprToTemp(E), type);
```
I don't think this code should be necessary?  E shouldn't be an lvalue or a record type, so we should just fall through to the same code at the end of the function, I think?

llvm-beanz replied:
Ah, yea, I think I need to fix this a different way.

The case where this happens is:

```c++
void fn(float x[2]);

void call(float Arr[2]) {
  fn(Arr);
}
```

In the CallExpr to fn, the argument is an lvalue DeclRefExpr to the ParmVar.

I think the right way to do this is to insert an ImplicitCastExpr for CK_HLSLArrayRValue or CK_LValueToRValue.

efriedma-quic commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -1126,6 +1126,7 @@ class ConstExprEmitter :
     case CK_NonAtomicToAtomic:
     case CK_NoOp:
     case CK_ConstructorConversion:
+    case CK_HLSLArrayRValue:
```
We shouldn't accept CK_HLSLArrayRValue casts here (or if we do, we need to do something different with them).

efriedma-quic commented on clang/lib/CodeGen/CGExprScalar.cpp:
```diff
@@ -2231,7 +2231,8 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
 
-  case CK_NoOp: {
+  case CK_NoOp:
+  case CK_HLSLArrayRValue: {
```
It shouldn't be possible to hit this case; a value of ArrayParameterType isn't a scalar.

efriedma-quic commented on clang/lib/CodeGen/CGExprComplex.cpp:
```diff
@@ -500,6 +500,7 @@ ComplexPairTy ComplexExprEmitter::EmitCast(CastKind CK, Expr *Op,
   case CK_NoOp:
   case CK_LValueToRValue:
   case CK_UserDefinedConversion:
+  case CK_HLSLArrayRValue:
```
The result of an CK_HLSLArrayRValue can't be a complex type.

efriedma-quic commented on clang/lib/CodeGen/CGExprAgg.cpp:
```diff
@@ -1442,6 +1445,7 @@ static bool castPreservesZero(const CastExpr *CE) {
   case CK_BitCast:
   case CK_ToUnion:
   case CK_ToVoid:
+  case CK_HLSLArrayRValue:
```
Probably should be false, like other lvalue-to-rvalue casts?

efriedma-quic reviewed (commented):
This seems like the right direction.

zygoloid commented on clang/include/clang/AST/Type.h:
```diff
@@ -3173,41 +3174,46 @@ class ArrayType : public Type, public llvm::FoldingSetNode {
 /// ConstantArrayType where the element type is 'int' and the size is 404.
-class ConstantArrayType final
-    : public ArrayType,
-      private llvm::TrailingObjects<ConstantArrayType, const Expr *> {
```
The size expression is usually not part of a constant array type. Normally, all we want as part of the type node is the bound, and the size expression if needed can be found on the `ArrayTypeLoc`. The rare special case being handled here is that if the array bound is instantiation-dependent but not dependent, then we need to preserve the expression in the type (not just in the `TypeLoc`) because it affects (for example) whether two template declarations involving such a type are considered to declare the same template. Because this is a very rare case, and `ConstantArrayType`s can be relatively common (eg, as the types of string literals), it seemed preferable to not store the `Expr*` if it's not needed, but in absolute terms it's probably a minor difference in memory usage, so it's probably fine to remove this optimization.

If we want to keep the optimization, perhaps we could replace the `Size` and `SizeExpr` fields with a pointer union that can hold either a <= 63-bit array bound (the common case) or a pointer to a slab-allocated pair of `APInt` and `const Expr*` (in the case where the bound doesn't fit in 63 bits or the size expression is instantiation-dependent).

llvm-beanz replied:
@zygoloid do you think the optimization is important to preserve? I guessed it might be safe to eliminate because other array types store the `SizeExpr` pointer, but if it's important I can make a separate PR to change the implementation as you suggested and we can merge that change first.

zygoloid replied:
As a general principle we try to keep our AST nodes small. This change isn't going to make a big difference by itself, but abandoning that principle and making a bunch of changes like this would. I think it's worth putting in a small amount of effort here, but if it doesn't work out nicely then it's fine to remove the optimization. We can find other places to save space.

llvm-beanz replied:
My team is trying to be pedantic about tracking our work, so I've filed #85124 to track implementing @zygoloid's suggestion. I'll try and get a PR up for that in the next day or two then update this PR on top of that change.

llvm-beanz committed:
Fix a logic error in my rebase

llvm-beanz committed:
Updates based on PR feedback

I beleive this captures all of the feedback from @efriedma-quic. Thank
you for your patience and great feedback!

efriedma-quic commented on clang/lib/AST/ASTContext.cpp:
```diff
@@ -10926,6 +10965,10 @@ QualType ASTContext::mergeTypes(QualType LHS, QualType RHS, bool OfBlockPointer,
+  case Type::ArrayParameter:
+    assert(LHS != RHS &&
+           "Equivalent pipe types should have already been handled!");
+    return LHS;
```
`return {};`?  Also, assertion text is wrong.

You could try to merge like we do for ConstantArray... but it's unlikely to matter, since HLSL is a C++ variant (and type merging is basically a C-only thing).

efriedma-quic commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -2331,6 +2332,11 @@ void CodeGenFunction::EmitVariablyModifiedType(QualType type) {
       type = cast<DecayedType>(ty)->getPointeeType();
       break;
 
+    case Type::ArrayParameter:
```
Can we just use the existing "case" for ConstantArray?

llvm-beanz committed:
Updates based on feedback

llvm-beanz committed:
clang-format

efriedma-quic reviewed (approved):
LGTM

coopp reviewed (approved):
WOW.. This was pretty big.  I went through it twice and didn't see anything that jumped out.  Looks good to me.

llvm-beanz committed:
Fix build kite

llvm-beanz committed:
Fix doc build warning

hokein commented on clang/test/SemaHLSL/ArrayTemporary.ll:
```diff
@@ -0,0 +1,76 @@
+; ModuleID = '/Users/cbieneman/dev/llvm-project/clang/test/SemaHLSL/ArrayTemporary.hlsl'
```
Is this `.ll` file necessary? It doesn't look like a lit test file, no `RUN` commands etc.
